### PR TITLE
Export managed artifact integrity diagnostics

### PR DESCRIPTION
**Pull request by Zigfreidish**
## Summary
- Add a model-ops export projection that surfaces persisted managed download `artifact_integrity` receipts as `managed_artifact_integrity_receipts[]`.
- Include strict-refusal diagnostics fields in `ExportResults` bundles: `verification_mode`, `policy_present`, `digest`, `checked_at`, and `failure_reason`.
- Document the export-bundle diagnostics contract in the Phase 8 local install runbook.

## Plan or Spec
- Governing plan: `docs/plans/2026-06-21-issue-1258-strict-install-diagnostics-bundle.md`
- Tracking issue: #1258
- Existing strict preflight context: `docs/plans/2026-05-25-issue-1258-strict-install-preflight.md`
- Runbook updated: `docs/runbooks/phase-8-local-install.md`

## Commands Run
```text
PYTHONPATH="$PWD:$PWD/services/mlx-worker-python" UV_PYTHON=3.12 UV_CACHE_DIR="$PWD/.uv-cache" uv run --project services/mlx-worker-python --extra mlx pytest services/mlx-worker-python/tests/test_benchmark_export.py -q
# 88 passed

PYTHONPATH="$PWD:$PWD/services/mlx-worker-python" UV_PYTHON=3.12 UV_CACHE_DIR="$PWD/.uv-cache" uv run --project services/mlx-worker-python pytest -q services/mlx-worker-python/tests/test_benchmark_export.py services/mlx-worker-python/tests/test_pr_scoped_performance.py::test_scope_report_selects_benchmark_export_probe services/mlx-worker-python/tests/test_pr_scoped_performance.py::test_registered_probes_expose_focused_commands services/mlx-worker-python/tests/test_pr_scoped_performance.py::test_probe_smokes_return_metrics_against_current_repo services/mlx-worker-python/tests/test_pr_scoped_performance.py::test_dispatch_probe_impl_supports_benchmark_export_probe services/mlx-worker-python/tests/test_pr_scoped_performance.py::test_probe_benchmark_export_run_scan_rejects_unexpected_job_count services/mlx-worker-python/tests/test_pr_scoped_performance.py::test_probe_benchmark_export_run_scan_rejects_unexpected_evaluation_job_count services/mlx-worker-python/tests/test_pr_scoped_performance.py::test_probe_benchmark_export_run_scan_rejects_unexpected_result_count services/mlx-worker-python/tests/test_pr_scoped_performance.py::test_probe_benchmark_export_run_scan_rejects_unexpected_summary_csv_count services/mlx-worker-python/tests/test_pr_scoped_performance.py::test_dispatch_probe_impl_supports_pr_scoped_scope_matcher_probe
# 97 passed via registered coverage run

git diff --check
# passed

PYTHONPATH="$PWD:$PWD/services/mlx-worker-python" UV_PYTHON=3.12 UV_CACHE_DIR="$PWD/.uv-cache" uv run --project services/mlx-worker-python python scripts/pr_scoped_performance_run.py --registry infra/perf/pr_scoped_probes.json --probe-id benchmark-export-run-scan-single-pass --base-repo /Users/chenyu/Documents/github/melix/.runtime/worktrees/perf-base-issue-1258-diagnostics --head-repo "$PWD" --output .runtime/perf/issue-1258-diagnostics-benchmark-export-rerun.json
# passed

git fetch origin main && git merge origin/main
# merged origin/main at 1debd8d8 into the task branch

PYTHONPATH="$PWD:$PWD/services/mlx-worker-python" UV_PYTHON=3.12 UV_CACHE_DIR="$PWD/.uv-cache" uv run --project services/mlx-worker-python --extra mlx pytest services/mlx-worker-python/tests/test_benchmark_export.py -q
# 88 passed after merging origin/main

PYTHONPATH="$PWD:$PWD/services/mlx-worker-python" UV_PYTHON=3.12 UV_CACHE_DIR="$PWD/.uv-cache" uv run --project services/mlx-worker-python --extra mlx pytest services/mlx-worker-python/tests/test_benchmark_export.py -q
# 91 passed after review-thread robustness fixes

PYTHONPATH="$PWD:$PWD/services/mlx-worker-python" UV_PYTHON=3.12 UV_CACHE_DIR="$PWD/.uv-cache" uv run --project services/mlx-worker-python coverage run -m pytest -q services/mlx-worker-python/tests/test_benchmark_export.py services/mlx-worker-python/tests/test_pr_scoped_performance.py::test_scope_report_selects_benchmark_export_probe services/mlx-worker-python/tests/test_pr_scoped_performance.py::test_registered_probes_expose_focused_commands services/mlx-worker-python/tests/test_pr_scoped_performance.py::test_probe_smokes_return_metrics_against_current_repo services/mlx-worker-python/tests/test_pr_scoped_performance.py::test_dispatch_probe_impl_supports_benchmark_export_probe services/mlx-worker-python/tests/test_pr_scoped_performance.py::test_probe_benchmark_export_run_scan_rejects_unexpected_job_count services/mlx-worker-python/tests/test_pr_scoped_performance.py::test_probe_benchmark_export_run_scan_rejects_unexpected_evaluation_job_count services/mlx-worker-python/tests/test_pr_scoped_performance.py::test_probe_benchmark_export_run_scan_rejects_unexpected_result_count services/mlx-worker-python/tests/test_pr_scoped_performance.py::test_probe_benchmark_export_run_scan_rejects_unexpected_summary_csv_count services/mlx-worker-python/tests/test_pr_scoped_performance.py::test_dispatch_probe_impl_supports_pr_scoped_scope_matcher_probe && PYTHONPATH="$PWD:$PWD/services/mlx-worker-python" UV_PYTHON=3.12 UV_CACHE_DIR="$PWD/.uv-cache" uv run --project services/mlx-worker-python coverage json -o coverage.json && python3 scripts/changed_scope_coverage.py --coverage-json coverage.json services/mlx-worker-python/worker/productization/benchmark_export.py services/mlx-worker-python/worker/productization/pr_scoped_performance.py services/mlx-worker-python/tests/test_benchmark_export.py services/mlx-worker-python/tests/test_pr_scoped_performance.py
# 100 passed; TOTAL changed-line coverage 98%

git diff --check origin/main...HEAD
# passed after merging origin/main

PYTHONPATH="$PWD:$PWD/services/mlx-worker-python" UV_PYTHON=3.12 UV_CACHE_DIR="$PWD/.uv-cache" uv run --project services/mlx-worker-python python scripts/pr_scoped_performance_run.py --registry infra/perf/pr_scoped_probes.json --probe-id benchmark-export-run-scan-single-pass --base-repo .runtime/perf-base-issue-1258-diagnostics-main --head-repo "$PWD" --output .runtime/perf/issue-1258-diagnostics-benchmark-export-after-merge.json
# passed after merging origin/main

PYTHONPATH="$PWD:$PWD/services/mlx-worker-python" UV_PYTHON=3.12 UV_CACHE_DIR="$PWD/.uv-cache" uv run --project services/mlx-worker-python python scripts/pr_scoped_performance_run.py --registry infra/perf/pr_scoped_probes.json --probe-id benchmark-export-run-scan-single-pass --base-repo .runtime/perf-base-issue-1258-review-fix --head-repo "$PWD" --output .runtime/perf/issue-1258-diagnostics-review-fix.json
# passed after review-thread robustness fixes

make swift-test
# passed in pre-commit

make py-test
# 4134 passed, 14 skipped

make integration-test
# 122 passed, 1 skipped

# Latest-main merge refresh after merging origin/main b9b9324e:
PYTHONPATH="$PWD:$PWD/services/mlx-worker-python" UV_PYTHON=3.12 UV_CACHE_DIR="$PWD/.uv-cache" uv run --project services/mlx-worker-python --extra mlx pytest -q services/mlx-worker-python/tests/test_benchmark_export.py services/mlx-worker-python/tests/test_pr_scoped_performance.py::test_scope_report_selects_benchmark_export_probe services/mlx-worker-python/tests/test_pr_scoped_performance.py::test_registered_probes_expose_focused_commands services/mlx-worker-python/tests/test_pr_scoped_performance.py::test_probe_smokes_return_metrics_against_current_repo services/mlx-worker-python/tests/test_pr_scoped_performance.py::test_dispatch_probe_impl_supports_benchmark_export_probe services/mlx-worker-python/tests/test_pr_scoped_performance.py::test_probe_benchmark_export_run_scan_rejects_unexpected_job_count services/mlx-worker-python/tests/test_pr_scoped_performance.py::test_probe_benchmark_export_run_scan_rejects_unexpected_evaluation_job_count services/mlx-worker-python/tests/test_pr_scoped_performance.py::test_probe_benchmark_export_run_scan_rejects_unexpected_result_count services/mlx-worker-python/tests/test_pr_scoped_performance.py::test_probe_benchmark_export_run_scan_rejects_unexpected_summary_csv_count services/mlx-worker-python/tests/test_pr_scoped_performance.py::test_dispatch_probe_impl_supports_pr_scoped_scope_matcher_probe
# 100 passed in 10.49s.

PYTHONPATH="$PWD:$PWD/services/mlx-worker-python" UV_PYTHON=3.12 UV_CACHE_DIR="$PWD/.uv-cache" uv run --project services/mlx-worker-python python3 - <<'PY'
from pathlib import Path
from scripts.pre_commit_gate import run_performance_report
changed_files = [
    "docs/plans/2026-06-21-issue-1258-strict-install-diagnostics-bundle.md",
    "docs/runbooks/phase-8-local-install.md",
    "services/mlx-worker-python/tests/test_benchmark_export.py",
    "services/mlx-worker-python/worker/productization/benchmark_export.py",
]
outcome = run_performance_report(Path.cwd(), changed_files, base_ref="origin/main")
print(f"report_dir={outcome.report_dir}")
print(f"status={outcome.status}")
print(f"selected_probe_count={outcome.selected_probe_count}")
if outcome.status != "ok":
    raise SystemExit(1)
PY
# status ok; selected probes 1; regressions 0; verification failures 0.
# Report: .runtime/pre-commit-performance/20260621-181735-a7cc1fcb/report/report.md.

# Final manual full gate after the merge commit was amended to ccf667d8.
# The amend hook skipped because there were no staged files, so the full gate
# was run manually.
make swift-test
# passed; macOS menubar package completed rc=0 elapsed=183s; final stage 831 tests in 25 suites passed.

make py-test
# 4165 passed, 14 skipped, 2 warnings in 176.91s.

make integration-test
# 122 passed, 1 skipped in 694.24s.

PYTHONPATH="$PWD:$PWD/services/mlx-worker-python" UV_PYTHON=3.12 UV_CACHE_DIR="$PWD/.uv-cache" uv run --project services/mlx-worker-python --extra mlx python -c 'from pathlib import Path
from scripts.pre_commit_gate import run_performance_report
changed_files = ["docs/plans/2026-06-21-issue-1258-strict-install-diagnostics-bundle.md", "docs/runbooks/phase-8-local-install.md", "services/mlx-worker-python/tests/test_benchmark_export.py", "services/mlx-worker-python/worker/productization/benchmark_export.py"]
outcome = run_performance_report(Path.cwd(), changed_files, base_ref="origin/main")
print(f"report_dir={outcome.report_dir}")
print(f"status={outcome.status}")
print(f"selected_probe_count={outcome.selected_probe_count}")
raise SystemExit(0 if outcome.status == "ok" else 1)
'
# status ok; selected probes 1; regressions 0; verification failures 0.
# Report: .runtime/pre-commit-performance/20260621-184435-ccf667d8/report/report.md.
```

## Coverage and Metrics
- Registered changed-scope coverage for `benchmark-export-run-scan-single-pass`: 98% total changed-line coverage; `benchmark_export.py` changed-line coverage 95.95%.
- PR-scoped performance report: `benchmark-export-run-scan-single-pass` passed in manual rerun, then passed again after merging current `origin/main`.
- Performance metrics: base `elapsed_ms_mean=56.876625`, head `elapsed_ms_mean=54.910258`; base `per_run_ms_mean=0.236986`, head `per_run_ms_mean=0.228793`; base `csv_elapsed_ms_mean=0.682617`, head `csv_elapsed_ms_mean=0.689809`.
- After-merge performance metrics: base `elapsed_ms_mean=54.632584`, head `elapsed_ms_mean=53.443992`; base `per_run_ms_mean=0.227636`, head `per_run_ms_mean=0.222683`; base `csv_elapsed_ms_mean=0.670525`, head `csv_elapsed_ms_mean=0.662858`.
- Review-fix performance metrics: base `elapsed_ms_mean=66.000867`, head `elapsed_ms_mean=66.150883`; base `per_run_ms_mean=0.275004`, head `per_run_ms_mean=0.275629`; base `csv_elapsed_ms_mean=0.718058`, head `csv_elapsed_ms_mean=0.742067`.
- Pre-commit performance analysis: the full hook reported one CSV-only threshold hit (`csv_elapsed_ms_mean` base `0.668`, head `0.704`, +`0.036ms`) while the export bundle metrics stayed neutral (`elapsed_ms_mean` +1.83%). A same-probe rerun produced base `csv_elapsed_ms_mean=0.736217`, head `0.674800`, so this was treated as measurement noise outside the changed path.
- Latest-main merge refresh scoped performance report: `.runtime/pre-commit-performance/20260621-184435-ccf667d8/report/report.md`
- Latest-main merge refresh coverage: `benchmark-export-run-scan-single-pass` passed with 98.0% changed-line coverage.
- Latest-main merge refresh metrics: base `elapsed_ms_mean=53.450`, head `elapsed_ms_mean=52.837`, delta `-1.15%`, neutral; base `per_run_ms_mean=0.223`, head `per_run_ms_mean=0.220`, delta `-1.15%`, neutral; base `csv_elapsed_ms_mean=0.656`, head `csv_elapsed_ms_mean=0.662`, delta `+0.92%`, neutral.
- Observability mode: evidence. The new data is a read-only export-bundle projection over existing local model-ops state files.
- Probe overhead: covered by the registered benchmark export run-scan probe; no new runtime download hot-path probe was added because this slice does not alter artifact materialization.

## Known Gaps
- This is a focused #1258 slice and does not close the full tracking issue.
- Release signing, publish-token minimization, async installer UX, desktop surfaces, and a full shared install/upgrade/bootstrap verification boundary remain deferred.
- No protobuf or dependency changes.

## Evidence Checklist
- [x] Relevant plan or spec is identified.
- [x] Behavior changes are reflected in the relevant docs.
- [x] Protocol changes include regenerated generated artifacts.
- [x] Dependency changes include updated lockfiles.
- [x] Relevant tests were run.
- [x] A metrics report is included, or `N/A` is stated explicitly with the reason.
- [x] Observability mode, probe overhead, and any deferred debug or evidence work are recorded, or `N/A` is stated explicitly with the reason.
- [x] Deferred work and known gaps are stated explicitly.

### CODE REVIEW

File: services/mlx-worker-python/worker/productization/benchmark_export.py [hunk 1]
```diff
@@ -294,6 +294,35 @@ def collect_agent_reliability_artifacts(
...
+    scanned_entries: _ScannedDirectoryEntries | None = None,
+) -> dict[str, object]:
+    receipts: list[dict[str, object]] = []
+    seen_paths: set[Path] = set()
+    for model_ops_root, root_scan in _candidate_model_ops_roots(
+        Path(jobs_root),
+        scanned_entries=scanned_entries,
+    ):
+        for state_path in _iter_model_ops_state_files(model_ops_root, scanned_entries=root_scan):
+            resolved_state_path = state_path.resolve()
```
[the remark this answers is not in the document]
Comment (Zigfreidish) [reply services/mlx-worker-python/worker/productization/benchmark_export.py]: Fixed in 028bdf50. `collect_model_ops_artifacts()` now resolves each state file through a non-fatal helper and skips paths that raise `OSError` or `RuntimeError`; `test_collect_model_ops_artifacts_skips_unresolvable_state_paths` and the symlink-loop state test cover the failure modes.

File: services/mlx-worker-python/worker/productization/benchmark_export.py [hunk 2]
```diff
@@ -784,6 +815,115 @@ def _resolve_agent_reliability_artifact_root(
...
+        directory, provided_scan = stack.pop()
+        scan = provided_scan if provided_scan is not None else _scan_directory(directory)
+        if scan is None:
+            continue
+        for name in reversed(scan.dir_names):
+            stack.append((directory / name, None))
+        for name in scan.file_names:
+            if name == "download.state.json" or name.endswith(".state.json"):
+                paths.append(directory / name)
+    return tuple(sorted(paths))
```
[the remark this answers is not in the document]
Comment (Zigfreidish) [reply services/mlx-worker-python/worker/productization/benchmark_export.py:878]: Fixed in 028bdf50. `_iter_model_ops_state_files()` now tracks visited directories by resolved path and skips unresolvable directories, which prevents directory symlink cycles from growing the DFS stack. Added `test_iter_model_ops_state_files_skips_directory_cycles`.

**Review by Zigfreidish — COMMENTED** [review]
Code review — no issues found, this looks ready to merge.

**`_try_load_json_object` exception broadening**
The change from `except OSError` to `except (OSError, TypeError, json.JSONDecodeError)` is a necessary improvement. A malformed JSON file would have raised `JSONDecodeError` and propagated unhandled before this fix, breaking the entire export for a single bad state file. Catching `TypeError` also guards against edge cases where the JSON root is not a dict.

**`_iter_model_ops_state_files` cycle detection**
The iterative DFS with `visited: set[Path]` on resolved paths correctly handles directory symlink loops. Using `_try_resolve_path` (catching `OSError, RuntimeError`) means unresolvable paths are silently skipped rather than crashing the scan.

**Activation decision logic**
`_managed_artifact_activation_decision` maps cleanly to the three observable states: `allowed` (completed + passed), `blocked` (terminal failure or artifact failure), `pending` (everything else). The `policy_present` coercion (`isinstance(policy_present, bool) else False`) is a sensible defensive guard against state files that store the field as a truthy string.

**Deduplication via resolved paths**
Using `seen_paths: set[Path]` with resolved absolute paths in `collect_model_ops_artifacts` correctly handles the symlink case tested in `test_collect_model_ops_artifacts_dedupes_resolved_state_paths`.

The one unchecked box in the plan doc (`- [ ] Commit the focused slice and open a PR`) is just the tracking document; this PR is that commit. 98% total changed-line coverage across the touched files.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: services/mlx-worker-python/worker/productization/benchmark_export.py [hunk 3]
```diff
@@ -839,7 +994,7 @@ def _load_json_object(path: Path) -> dict[str, object]:
 def _try_load_json_object(path: Path) -> dict[str, object] | None:
     try:
         return _load_json_object(path)
-    except OSError:
+    except (OSError, TypeError, json.JSONDecodeError):
```
Comment (Zigfreidish) [inline services/mlx-worker-python/worker/productization/benchmark_export.py:997]: Good catch expanding `_try_load_json_object` to also swallow `TypeError` and `json.JSONDecodeError`. Previously a file whose bytes decoded cleanly but whose content wasn't a JSON object (e.g. a bare string or array) would surface a `TypeError` at the caller rather than silently being skipped. Now all three failure modes (I/O error, parse error, type mismatch) are handled uniformly.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: services/mlx-worker-python/worker/productization/benchmark_export.py
Comment (Zigfreidish) [inline services/mlx-worker-python/worker/productization/benchmark_export.py:1]: The `activation_decision` mapping is clean and complete. One observation: `stalled` is included in the `blocked` set, but `stalled` jobs may later transition to `failed` or be retried. Since this projection is read-only and snapshot-at-export-time, treating `stalled` as `blocked` is the conservative and correct choice for a diagnostics surface.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by Zigfreidish — COMMENTED** [review]
Solid read-only diagnostics projection. Path deduplication via `resolved_state_path`, symlink-loop protection in `_iter_model_ops_state_files`, and the non-fatal handling of malformed/unrelated state files are all done correctly. The `activation_decision` logic is clean and deterministic. Test matrix covers the skipping, deduplication, and activation-decision paths well. Aggregate changed-scope coverage 98%. LGTM — minor notes below are non-blocking.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by Zigfreidish — COMMENTED** [review]
**Review: LGTM — ready to merge**

Clean read-only projection. Key observations:

**`_try_load_json_object` fix is important.** The original only caught `OSError`; the new version also catches `TypeError` and `json.JSONDecodeError`, which is correct — malformed JSON is the expected failure mode for corrupted state files, not OS errors. `test_collect_model_ops_artifacts_skips_malformed_or_unrelated_state_files` covers the broken-JSON case.

**Symlink-loop and cycle detection is solid.** `_iter_model_ops_state_files` uses `visited: set[Path]` of resolved paths, which correctly deduplicates real files reached via different symlink paths (`test_collect_model_ops_artifacts_dedupes_resolved_state_paths`) and breaks directory symlink cycles (`test_iter_model_ops_state_files_skips_directory_cycles`). The unresolvable-path test via `monkeypatch` on `Path.resolve` covers the `OSError` case cleanly.

**`_managed_artifact_activation_decision` logic is correct** for all three cases: `completed`+`passed` → `allowed`, any terminal failure state or `failed` integrity → `blocked`, everything else → `pending`. The test validates all three branches.

**`policy_present` coercion** (`if isinstance(policy_present, bool) else False`) correctly rejects stringy truthy values like `"yes"` from older state files while preserving actual booleans. The test with `"policy_present": "yes"` confirming it maps to `False` is exactly the right coverage.

No issues found.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by Zigfreidish — COMMENTED** [review]
## Code Review: Export managed artifact integrity diagnostics

### Summary
Read-only export-bundle projection of managed artifact integrity receipts. No changes to the download hot path. Clean and well-tested.

---

### `benchmark_export.py`

**`collect_model_ops_artifacts`**
- Deduplicates by resolved path — correctly handles symlinked state files that point to the same underlying file.
- Scans both the jobs root (if it looks like a model-ops dir) and an explicit `model-ops/` subdirectory, with separate `seen` tracking for roots to avoid double-scanning.

**`_iter_model_ops_state_files`**
- Explicit stack + visited-set instead of `os.walk` correctly prevents symlink-cycle hangs. Tested explicitly by `test_iter_model_ops_state_files_skips_directory_cycles`.

**`_try_load_json_object` change**
```python
except (OSError, TypeError, json.JSONDecodeError):
```
Previously only caught `OSError`. The expanded exception list is the right fix — malformed state files (truncated JSON, binary blobs) should be non-fatal at the collector layer. Tested by `test_collect_model_ops_artifacts_skips_malformed_or_unrelated_state_files`.

**`_managed_artifact_activation_decision`**
- `"allowed"` only when both terminal state is `completed` AND artifact status is `passed` — correct strict semantics.
- `"blocked"` for any failure/cancel/stall or failed artifact — correctly fail-closed.
- `"pending"` for everything else (in-progress, pending, unknown states).

**`policy_present` handling**
```python
"policy_present": policy_present if isinstance(policy_present, bool) else False,
```
Correct: state files written with `"yes"` or `1` coerce to `False` rather than passing through as a non-bool, keeping the export schema type-stable.

---

### Tests

Well-covered edge cases:
- Malformed JSON, non-download operations, non-dict `artifact_integrity`, wrong operation type.
- Symlink deduplication.
- Unresolvable paths (monkeypatched `Path.resolve`).
- Symlink loop to self.
- Disappearing directories during scan.
- `_has_model_ops_state_markers(None) is False` — tested explicitly.

---

### Minor Observation
`_has_model_ops_state_markers` matches `name.endswith(".state.json")` broadly. Files like `foo.state.json` under a model-ops root will be scanned even if they aren't download manifests; `_managed_artifact_integrity_receipt_row` correctly returns `None` for them (wrong `operation` or missing `artifact_integrity`), so there is no correctness issue — just a slight over-scan at the file-discovery layer.

---

### Verdict
Clean, well-tested, zero hot-path impact. Would approve if self-approval were permitted.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: services/mlx-worker-python/worker/productization/benchmark_export.py
Comment (Zigfreidish) [inline services/mlx-worker-python/worker/productization/benchmark_export.py:1]: The broadening of `_try_load_json_object` to catch `(OSError, TypeError, json.JSONDecodeError)` in addition to just `OSError` is the right call — state files can be partially written or contain truncated JSON, and silently skipping them is the correct non-fatal policy for a read-only diagnostics projection. Good to pair this with the dedicated `test_collect_model_ops_artifacts_skips_malformed_or_unrelated_state_files` test.

---
_Generated by [Claude Code](https://claude.ai/code)_

Comment (Zigfreidish) [inline services/mlx-worker-python/worker/productization/benchmark_export.py:1]: `_iter_model_ops_state_files` uses an explicit `visited: set[Path]` of resolved paths to break symlink cycles, matching the pattern used elsewhere in the codebase for directory traversal. The `_try_resolve_path` helper returning `None` on `OSError | RuntimeError` keeps the loop non-fatal for any unresolvable path. The stack-based traversal (rather than `os.walk`) is the right choice here since `os.walk` does not protect against symlink loops by default.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by Zigfreidish — COMMENTED** [review]
The read-only export projection is well-scoped: it touches no download hot paths and only reads persisted state files at export time. The `collect_model_ops_artifacts` implementation handles all the tricky filesystem edge cases — symlink deduplication via resolved paths, directory cycles via the visited set, partial/malformed JSON, disappearing directories, and symlink loops — with dedicated tests for each. The `activation_decision` derivation (`allowed` / `blocked` / `pending`) is straightforward and correctly normalises non-bool `policy_present` values. 98% changed-scope coverage and a full pre-commit gate pass. **LGTM — ready to merge.**

---
_Generated by [Claude Code](https://claude.ai/code)_

File: services/mlx-worker-python/worker/productization/benchmark_export.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Zigfreidish) [reply services/mlx-worker-python/worker/productization/benchmark_export.py]: Fixed earlier in the review follow-up: resolution now goes through `_try_resolve_path`, which catches `OSError` and `RuntimeError` and skips unreadable or unresolvable state roots. The latest focused suite and scoped performance run on top of origin/main b9b9324e passed.

File: services/mlx-worker-python/worker/productization/benchmark_export.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Zigfreidish) [reply services/mlx-worker-python/worker/productization/benchmark_export.py:878]: Fixed earlier in the review follow-up: `_iter_model_ops_state_files` now tracks a `visited` set of resolved directories and skips unresolvable paths, so symlink cycles do not recurse indefinitely. Covered by `test_iter_model_ops_state_files_skips_directory_cycles` and rerun in the latest focused suite.
